analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs

```diff
@@ -0,0 +1,149 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#if NET
+
+using SonarAnalyzer.Rules.CSharp;
+
+namespace SonarAnalyzer.Test.Rules;
+
+[TestClass]
+public class CallModelStateIsValidTest
+{
+    private readonly VerifierBuilder builder = new VerifierBuilder<CallModelStateIsValid>()
+        .WithBasePath("AspNet")
+        .AddReferences([
+            AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcAbstractions,
+            AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcCore,
+            AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcViewFeatures,
+            ..NuGetMetadataReference.SystemComponentModelAnnotations(Constants.NuGetLatestVersion)
+        ]);
+
+    [TestMethod]
+    public void CallModelStateIsValid_CS() =>
+        builder.AddPaths("CallModelStateIsValid.cs").Verify();
+
+    [TestMethod]
+    public void CallModelStateIsValid_AssemblyLevelControllerAttribute_CS() =>
+        builder.AddSnippet("""
+            using Microsoft.AspNetCore.Mvc;
+            using System.ComponentModel.DataAnnotations;
+
+            [assembly: ApiController]           // causes the ApiController attribute to be applied on every Controller class in the assembly
+
+            public class Movie
+            {
+                [Required] public string Title { get; set; }
+                [Range(1900, 2200)] public int Year { get; set; }
+            }
+
+            public class ControllerWithApiAttributeAtTheAssemblyLevel : ControllerBase
+            {
+                [HttpPost("/[controller]")]
+                public string Add(Movie movie)  // Compliant - ApiController attribute is applied at the assembly level.
+                {
+                    return "Hello!";
+                }
+            }
+            """).Verify();
+
+    [TestMethod]
+    public void CallModelStateIsValid_FluentValidation_CS() =>
+        builder.AddSnippet("""
+            using Microsoft.AspNetCore.Mvc;
+            using FluentValidation;
+
+            public class Movie
+            {
+                public string Title { get; set; }
+                public int Year { get; set; }
+            }
+
+            public class MovieValidator : AbstractValidator<Movie>
+            {
+              public MovieValidator()
+              {
+                RuleFor(x => x.Title).NotNull();
+                RuleFor(x => x.Year).InclusiveBetween(1900, 2100);
+              }
+            }
+
+            public class MovieController : ControllerBase
+            {
+                private IValidator<Movie> _validator;
+
+                public MovieController(IValidator<Movie> validator)
+                {
+                    _validator = validator;
+                }
+
+                [HttpPost("/[controller]")]
+                public string Add(Movie movie)  // Compliant - uses FluentValidation
+                {
+                    if (!_validator.Validate(movie).IsValid)
```

Comment:
Nice. (This pattern would probably also work for POCO controllers).

Reply:
I was referring to injecting the `IValidator` service. That way, you can validate in a POCO controller action. Please also take a look at this comment https://github.com/SonarSource/sonar-dotnet/pull/9183/files#r1579080060 which you fixed.